### PR TITLE
fix(controller): allow to use global params with withParam

### PR DESCRIPTION
**Pull request by dtaniwaki**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.

(b)

* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

### CODE REVIEW

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -526,7 +526,12 @@ func (woc *wfOperationCtx) resolveDependencyReferences(dagCtx *dagContext, task
 		return nil, errors.InternalWrapError(err)
 	}
 	fstTmpl := fasttemplate.New(string(taskBytes), "{{", "}}")
-	newTaskStr, err := common.Replace(fstTmpl, scope.replaceMap(), true)
+
+	replaceMap := scope.replaceMap()
```
Comment (simster7) [inline workflow/controller/dag.go:530]: I think something like 
```suggestion
	replaceMap := scope.replaceMap(woc.globalParams)
```

Would be cleaner.

You could edit the `replaceMap` function to include this optional parameter and add its contents to the scope if it's not nil

Comment (simster7) [reply workflow/controller/dag.go:530]: What do you think?

Comment (dtaniwaki) [reply workflow/controller/dag.go:530]: I think the method name `replaceMap` is not intuitive to accept an argument of params extension. Maybe, we can have something like `mergeParams` utility method in the common package.

Comment (dtaniwaki) [reply workflow/controller/dag.go:530]: If we want to introduce `mergeParams` utility method, we should make another PR to replace other codes meging params.

Comment (simster7) [reply workflow/controller/dag.go:547]: We could change the name to something like `prepareScope`, here is an example where I do something similar:

https://github.com/argoproj/argo/blob/7cb2fd17765aad691eda25ca4c5acecb89f84394/workflow/controller/steps.go#L466-L471

This is minor though, I'll leave it up to you.

Comment (dtaniwaki) [reply workflow/controller/dag.go]: As I found there're many cases of merging params, I defined `Params` type which has `Clone` and `Merge` functions so we can replace those codes with one line.

I also found the order of merging was not correct in my original commit of this PR. `scoep.replaceMap()` should be merged into `globalParmas` but not the opposite.

**Comment by dtaniwaki** [comment]
@simster7 Could you review it again?

File: util/params/params.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package params
```
Comment (simster7) [inline util/params/params.go:1]: This is a Workflow specific util, so it shouldn't be at the top-level util folder. Can you add this to the `workflow/common` package? You can add a new file there, but it should still be under the `common` package.

File: util/params/params.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+type Params map[string]string
```
Comment (simster7) [inline util/params/params.go:3]:
```suggestion
type Parameters map[string]string
```
Full name is more clear.

File: workflow/common/util.go [hunk 4]
```diff
@@ -300,19 +301,13 @@ func ProcessArgs(tmpl *wfv1.Template, args wfv1.ArgumentsProvider, globalParams,
 }
 
 // SubstituteParams returns a new copy of the template with global, pod, and input parameters substituted
-func SubstituteParams(tmpl *wfv1.Template, globalParams, localParams map[string]string) (*wfv1.Template, error) {
+func SubstituteParams(tmpl *wfv1.Template, globalParams, localParams params.Params) (*wfv1.Template, error) {
```
Comment (simster7) [inline workflow/common/util.go:304]: When we move the util file under `common`, we won't need the import prefix anymore.

File: util/params/params_test.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+	assert.Equal(t, Params{"foo": "2", "bar": "1", "wow": "1"}, newParams)
+	assert.NotSame(t, &params, &newParams)
+}
+
+// TestParamsClone ensures Clone of Params works correctly.
+func TestParamsClone(t *testing.T) {
+	params := Params{"foo": "1"}
+	newParams := params.Clone()
+	assert.Equal(t, params, newParams)
+	assert.NotSame(t, &params, &newParams)
```
Comment (simster7) [inline util/params/params_test.go:22]: Nice, this is important.

File: util/params/params.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+	newParams := ps.Clone()
+	for _, params := range args {
+		for k, v := range params {
+			newParams[k] = v
+		}
+	}
+	return newParams
+}
+
+func (ps Params) Clone() Params {
```
Comment (simster7) [inline util/params/params.go:15]: `DeepCopy` is the name we use for this in Argo.

File: workflow/controller/scope.go [hunk 7]
```diff
@@ -14,7 +15,7 @@ type wfScope struct {
 }
 
 // replaceMap returns a replacement map of strings intended to be used simple string substitution
-func (s *wfScope) replaceMap() map[string]string {
+func (s *wfScope) replaceMap() params.Params {
 	replaceMap := make(map[string]string)
```
Comment (simster7) [inline workflow/controller/scope.go:19]: Shouldn't this be?
```suggestion
	replaceMap := make(params.Params)
```

File: workflow/controller/scope.go [hunk 8]
```diff
@@ -14,8 +15,8 @@ type wfScope struct {
 }
 
 // replaceMap returns a replacement map of strings intended to be used simple string substitution
-func (s *wfScope) replaceMap() map[string]string {
-	replaceMap := make(map[string]string)
+func (s *wfScope) replaceMap() common.Parameters {
```
Comment (simster7) [inline workflow/controller/scope.go:18]: Please rename this function `parameters()` or `getParameters()` or something similar

File: workflow/controller/workflowpod.go [hunk 9]
```diff
@@ -314,11 +314,8 @@ func (woc *wfOperationCtx) createWorkflowPod(nodeName string, mainCtr apiv1.Cont
 }
 
 // substitutePodParams returns a pod spec with parameter references substituted as well as pod.name
-func substitutePodParams(pod *apiv1.Pod, globalParams map[string]string, tmpl *wfv1.Template) (*apiv1.Pod, error) {
-	podParams := make(map[string]string)
-	for k, v := range globalParams {
-		podParams[k] = v
-	}
+func substitutePodParams(pod *apiv1.Pod, globalParams common.Parameters, tmpl *wfv1.Template) (*apiv1.Pod, error) {
+	podParams := globalParams.Merge()
```
Comment (simster7) [inline workflow/controller/workflowpod.go:318]: Shouldn't this be `DeepCopy` if `Merge` has not arguments?
```suggestion
	podParams := globalParams.DeepCopy()
```

**Comment by dtaniwaki** [comment]
@simster7 Fixed. Please check it again!

**Comment by simster7** [comment]
Back-ported to 2.7

**Pull request merged**